src/coreclr/jit/decomposelongs.cpp

```diff
@@ -562,28 +570,78 @@ GenTree* DecomposeLongs::DecomposeStoreLclFld(LIR::Use& use)
 GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
 {
     assert(use.IsInitialized());
-    assert(use.Def()->OperGet() == GT_CAST);
+    assert(use.Def()->OperIs(GT_CAST));
 
-    GenTree* cast     = use.Def()->AsCast();
-    GenTree* loResult = nullptr;
-    GenTree* hiResult = nullptr;
+    GenTreeCast* cast    = use.Def()->AsCast();
+    var_types    srcType = cast->CastFromType();
+    var_types    dstType = cast->CastToType();
 
-    var_types srcType = cast->CastFromType();
-    var_types dstType = cast->CastToType();
-
-    if ((cast->gtFlags & GTF_UNSIGNED) != 0)
+    if (cast->IsUnsigned())
     {
         srcType = varTypeToUnsigned(srcType);
     }
 
-    bool skipDecomposition = false;
+#if defined(FEATURE_HW_INTRINSICS) && defined(TARGET_X86)
+    if (varTypeIsFloating(dstType))
+    {
+        // We will reach this path only if morph did not convert the cast to a helper call,
+        // meaning we can perform the cast using SIMD instructions.
+        // The sequence this creates is simply:
+        //    AVX512DQ.VL.ConvertToVector128Single(Vector128.CreateScalarUnsafe(LONG)).ToScalar()
+
+        NamedIntrinsic intrinsicId      = NI_Illegal;
+        GenTree*       srcOp            = cast->CastOp();
+        var_types      dstType          = cast->CastToType();
+        CorInfoType    baseFloatingType = (dstType == TYP_FLOAT) ? CORINFO_TYPE_FLOAT : CORINFO_TYPE_DOUBLE;
+        CorInfoType    baseIntegralType = cast->IsUnsigned() ? CORINFO_TYPE_ULONG : CORINFO_TYPE_LONG;
+
+        assert(!cast->gtOverflow());
+
+        if (m_compiler->compOpportunisticallyDependsOn(InstructionSet_AVX512DQ_VL))
+        {
+            intrinsicId = (dstType == TYP_FLOAT) ? NI_AVX512DQ_VL_ConvertToVector128Single
+                                                 : NI_AVX512DQ_VL_ConvertToVector128Double;
+        }
+        else
+        {
+            assert(m_compiler->compIsaSupportedDebugOnly(InstructionSet_AVX10v1));
+            intrinsicId =
+                (dstType == TYP_FLOAT) ? NI_AVX10v1_ConvertToVector128Single : NI_AVX10v1_ConvertToVector128Double;
+        }
```

Comment:
These checks feel like they should be inverted since `AVX10v1` is newer, so we should opportunistically check for `AVX10v1` and assume otherwise `AVX512DQ.VL` is supported. This is particularly relevant since the spec for AVX10 is changing to require V512 support.

Reply:
Actually, with the change to require V512, I think we don't even need the AVX10v1 path anymore, technically. As AVX512DQ.VL will always be available if AVX10v1 is available

Reply:
This is copying the pattern we already use in a lot of other places, but yeah, we can simplify all of them now, I think.